### PR TITLE
Use grpc for communication from worker to local raylet (task submission and direct actor args only)

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Eventually, we need workers to be able to contact remote raylets directly. Also, we should unify the communication between workers and raylets so that it all goes through grpc, instead of asio in one direction and grpc back.

This PR begins this effort by giving each worker a grpc client to the local raylet. It also ports two existing calls to use grpc only:
- SubmitTask
- WaitForDirectActorCallArgs

We should port more calls to grpc eventually, but for now, sticking with some simple calls that have no ordering requirements.

This also sets a default port (2080) for raylet so that in the case where the driver starts a cluster, it can find the right node manager to connect to.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/parameter.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ def __init__(self,
                  redis_port=None,
                  redis_shard_ports=None,
                  object_manager_port=None,
-                 node_manager_port=None,
+                 node_manager_port=ray_constants.DEFAULT_NODE_MANAGER_PORT,
```
Comment (ericl) [inline python/ray/parameter.py]: Wouldn't this break pretty often if you restart your console or are trying to run two Python programs at once?

File: src/ray/core_worker/core_worker.h [hunk 2]
```diff
@@ -25,10 +26,9 @@
 /// 1) Add the rpc to the WorkerService in core_worker.proto, e.g., "ExampleCall"
 /// 2) Add a new handler to the macro below: "RAY_CORE_WORKER_RPC_HANDLER(ExampleCall, 1)"
 /// 3) Add a method to the CoreWorker class below: "CoreWorker::HandleExampleCall"
-#define RAY_CORE_WORKER_RPC_HANDLERS                       \
-  RAY_CORE_WORKER_RPC_HANDLER(AssignTask, 5)               \
-  RAY_CORE_WORKER_RPC_HANDLER(DirectActorAssignTask, 9999) \
-  RAY_CORE_WORKER_RPC_HANDLER(DirectActorCallArgWaitComplete, 100)
+#define RAY_CORE_WORKER_RPC_HANDLERS         \
+  RAY_CORE_WORKER_RPC_HANDLER(AssignTask, 5) \
+  RAY_CORE_WORKER_RPC_HANDLER(DirectActorAssignTask, 9999)
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h]: I'm not sure removing the arg wait callback RPC is a good idea, unless we are fine with very large number of async RPCs to the node manager (like 100k+).

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 3]
```diff
@@ -244,12 +239,17 @@ class SchedulingQueue {
...
         RAY_CHECK(boost::this_thread::get_id() == main_thread_id_);
         auto it = pending_tasks_.find(seq_no);
         if (it != pending_tasks_.end()) {
           it->second.MarkDependenciesSatisfied();
-          ScheduleRequests();
+          if (!status.ok()) {
+            it->second.Cancel();
+            pending_tasks_.erase(it);
+          } else {
+            ScheduleRequests();
```
Comment (ericl) [inline src/ray/core_worker/transport/direct_actor_transport.h]: Can we just call ScheduleRequests() in all cases? The requests will fail by themselves.

I'm pretty sure the cancellation path above is also incomplete.

File: src/ray/rpc/node_manager/node_manager_server.h [hunk 4]
```diff
@@ -70,6 +92,10 @@ class NodeManagerGrpcService : public GrpcService {
             main_service_));
 
     // Set accept concurrency.
+    server_call_factories_and_concurrencies->emplace_back(
+        std::move(submit_task_call_factory), 100);
+    server_call_factories_and_concurrencies->emplace_back(
+        std::move(wait_for_direct_actor_call_args_factory), 100);
```
Comment (ericl) [inline src/ray/rpc/node_manager/node_manager_server.h]: The concurrency should be much higher for this if we are going with this strategy.

File: src/ray/core_worker/test/scheduling_queue_test.cc [hunk 5]
```diff
@@ -64,6 +64,37 @@ TEST(SchedulingQueueTest, TestWaitForObjects) {
...
+  ASSERT_EQ(n_ok, 0);
+  ASSERT_EQ(n_rej, 0);
+
+  waiter.Complete(2, Status::Invalid(""));
+  ASSERT_EQ(n_ok, 0);
+  ASSERT_EQ(n_rej, 1);
+
+  waiter.Complete(0);
+  ASSERT_EQ(n_ok, 2);
+}
```
Comment (ericl) [inline src/ray/core_worker/test/scheduling_queue_test.cc]: I'm scared of this error handling path. Can we avoid adding a new edge case and just treat failed dependency waits as success?

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 3 above)
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.h]: How would they fail by themselves? I think we still need to cancel individual tasks. But yeah, I'll put the ScheduleRequests() call back and add it to the unit test.

Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: They will execute, call ray.get() on the arg, and that will fail. There is no need to cancel them.

Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Oh I see, the task will just fail if the arguments aren't local? I like that better too, will revert.

File: python/ray/parameter.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/parameter.py]: If you restart your console, then ray should also get restarted, but yeah it would in the latter case. Then how about for drivers that start ray, we can first pick an unused port from the driver side?

Comment (ericl) [reply python/ray/parameter.py]: I find a lot of times the port doesn't get released for a while, which is the reason you sometimes need to pick a random port.

It might possibly work to pick an unused _random_ port on the driver side (there is a slight race condition that someone claims it before the raylet starts, but this is unlikely so probably acceptable).

File: python/ray/node.py [hunk 6]
```diff
@@ -390,6 +406,13 @@ def new_log_files(self, name, redirect_output=True):
         log_stderr_file = open(log_stderr, "a", buffering=1)
         return log_stdout_file, log_stderr_file
 
+    def _get_unused_port(self):
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.bind(("", 0))
+        port = s.getsockname()[1]
+        s.close()
```
Comment (ericl) [inline python/ray/node.py]: nice

File: src/ray/rpc/node_manager/node_manager_client.h [hunk 7]
```diff
@@ -57,6 +57,53 @@ class NodeManagerClient {
   ClientCallManager &client_call_manager_;
 };
 
+/// Client used by workers for communicating with a node manager server.
+class NodeManagerWorkerClient
```
Comment (ericl) [inline src/ray/rpc/node_manager/node_manager_client.h]: How about just `NodeManagerClient`?

File: src/ray/raylet/raylet_client.cc [hunk 8]
```diff
@@ -203,8 +203,13 @@ ray::Status RayletConnection::AtomicRequestReply(
 
 RayletClient::RayletClient(const std::string &raylet_socket, const WorkerID &worker_id,
                            bool is_worker, const JobID &job_id, const Language &language,
+                           std::shared_ptr<ray::rpc::NodeManagerWorkerClient> grpc_client,
```
Comment (ericl) [inline src/ray/raylet/raylet_client.cc]: Consider moving the node manager client into a top level field of core_worker.h

**Review by ericl — APPROVED** [review]
LGTM.

File: src/ray/rpc/node_manager/node_manager_client.h (hunk 7 above)
Comment (stephanie-wang) [reply src/ray/rpc/node_manager/node_manager_client.h]: The other one for cross-NodeManager communication is called that :/ or are you suggesting to combine?

File: python/ray/node.py [hunk 9]
```diff
@@ -144,6 +148,16 @@ def __init__(self,
...
 
+        if head or not connect_only:
+            # We need to start a local raylet.
+            if (self._ray_params.node_manager_port is None
+                    or self._ray_params.node_manager_port == 0):
+                # No port specified. Pick a random port for the raylet to use.
+                # NOTE: There is a possible but unlikely race condition where
+                # the port is bound by another process between now and when the
+                # raylet starts.
+                self._ray_params.node_manager_port = self._get_unused_port()
```
Comment (zhijunfu) [inline python/ray/node.py]: Currently worker talks to raylet via UNIX domain socket, and grpc supports that too, that UNIX socket name should not conflict with other applications. Do we want to use TCP socket for local worker connection also? (which is fine, but just want to confirm). 

And if the intent is to let the driver know the port of the raylet, maybe we could instead let driver do a sync read to get the raylet port from GCS, as it needs to connect to GCS anyway.

**Review by zhijunfu — COMMENTED** [review]
Thanks, it's nice to migrate the flatbuf stuff to grpc. 

Looks good overall, I left a comment on the node manager port.

File: python/ray/node.py (hunk 9 above)
Comment (stephanie-wang) [reply python/ray/node.py]: Yes, we're using TCP just to make all worker-raylet connections uniform. Eventually, we would like to submit tasks directly from a worker to a remote raylet too.

Hmm I thought about doing it with sync reads to the GCS, but decided against it. For testing purposes, we often connect multiple raylets to the GCS on a single node, so we'd have to differentiate between them to find the right raylet. It's a bit of a headache.

Comment (zhijunfu) [reply python/ray/node.py]: @stephanie-wang How about we choose a random node ID instead and pass that to the driver? then driver can wait for this node ID to appear on GCS and get the port for the specific raylet then. There would be no collisions in this case.

Comment (stephanie-wang) [reply python/ray/node.py]: I might do that for #6143, but I'll hold off on getting the raylet port from the GCS. I don't think it's a big deal to set the port from the driver for now.
